src/solvers/refinement/string_refinement.cpp

```diff
@@ -1085,21 +1095,42 @@ static exprt negation_of_constraint(const string_constraintt &axiom)
   return negaxiom;
 }
 
+/// Result of the solver `supert` should not be interpreted literally for char
+/// arrays as not all indices are present in the index set.
+/// In the given expression, we populate arrays at the indices for which the
+/// solver has no constraint by copying values to the left.
+/// For example an expression `ARRAY_OF(0) WITH [1:=2] WITH [4:=3]` would
+/// be interpreted as `{ 2, 2, 3, 3, 3}`.
+/// \param expr: expression to interpret
+/// \param string_max_length: maximum size of arrays to consider
+/// \return the interpreted expression
+exprt concretize_arrays_in_expression(exprt expr, std::size_t string_max_length)
+{
+  for(auto op=expr.depth_begin(); op!=expr.depth_end();)
+  {
+    if(op->id()==ID_with && op->type().id()==ID_array)
+    {
+      op.mutate()=fill_in_array_with_expr(*op, string_max_length);
+      op.next_sibling_or_parent();
```

Comment:
I was quite suspicious about that thing but I now believe it's OK on the basis that we can safely interrupt the depth-first search when we find an array, as we should not find nested arrays after the solver run (even though we don't really check that).